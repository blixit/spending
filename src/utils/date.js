/**
 * Formats a date for the backend
 * @param {Date} date the date to convert
 */
export const toBackendDate = (date) => {
  return date
    ? date.toLocaleDateString() + ' ' + date.toLocaleTimeString()
    : date;
};