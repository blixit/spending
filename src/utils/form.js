/**
* Removes the listed fields from the source
* @param {*} form The source object
* @param {*} fields The fields list
*/
export const clean = (form, fields) => {
  const result = {};
  // eslint-disable-next-line no-unused-vars
  for (const field in form) {
    if (!form.hasOwnProperty(field) || fields.includes(field)) {
      continue;
    }
    result[field] = form[field];
  }
  return result;
};
