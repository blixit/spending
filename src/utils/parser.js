export const stringify = data => {
  try {
    return JSON.stringify(data);
  } catch (error) {
    return null;
  };
}

export const parse = data => JSON.parse(data);