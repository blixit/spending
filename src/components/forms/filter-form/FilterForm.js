import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

import { mutate, ReadyMutations as Mutations } from 'core/http/query';
import { toBackendDate } from 'utils/date';

import ThemedButton from 'ui/atoms/buttons/ThemedButton';

import StyledError from 'components/molecules/error/Error';
import FormInput from 'ui/forms/form-input/FormInput';
import FormDate from 'ui/forms/form-date/FormDate';
import FormSelector from 'ui/forms/form-selector/FormSelector';

const Form = styled.form`
  display: flex;
  flex-direction: column;
`;

const FormNumberInput = styled(FormInput)`
  max-width: 150px;
`;

const Block = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  flex: auto;
`;

class FilterForm extends React.Component {
  static propTypes = {
    onFormResult: PropTypes.func
  };

  categories = [
    { id: 1, option: 'Restauration' },
    { id: 2, option: 'Taxis' }
  ];

  constructor (props) {
    super(props);
    const now = new Date();
    const before = new Date(new Date().setMonth(now.getMonth() - 1));
    this.state = {
      category: '',
      dateStart: before,
      dateEnd: now,
      minPrice: undefined,
      maxPrice: undefined,
      name: undefined,
      hasError: false,
      errors: {}
    };
  }

  onCategorySelected = e => this.setState({ category: e.target.value });

  onDateStartChanged = dateStart => this.setState({ dateStart });

  onDateEndChanged = dateEnd => this.setState({ dateEnd });

  onPriceMinChanged = minPrice => this.setState({ minPrice });

  onPriceMaxChanged = maxPrice => this.setState({ maxPrice });
  
  onNameChanged = name => this.setState({ name });

  onSubmit = async e => {
    e.preventDefault();

    const { errors, dateStart, dateEnd, ...rest } = this.state;
    const { onFormResult } = this.props;
    const { search } = Mutations.spending;

    const body = {
      ...rest,
      dateStart: toBackendDate(dateStart),
      dateEnd: toBackendDate(dateEnd)
    };

    try {
      const response = await mutate({ ...search, data: body });
      onFormResult && onFormResult(response);
      this.setState({ hasError: false });
    } catch (e) {
      errors.global = search.queryError || e.message;
      this.setState({ hasError: true, errors });
    }
  };

  render() {
    const { dateStart, dateEnd, minPrice, maxPrice, name, hasError, errors } = this.state;
    const { onFormResult, ...rest } = this.props;

    return (
      <Form {...rest}>
        <Block>
          <FormDate
            label='From'
            selected={dateStart}
            onChange={this.onDateStartChanged}
          />
          <FormDate arabic
            label='To'
            selected={dateEnd}
            onChange={this.onDateEndChanged}
          />
        </Block>
        <Block>
          <FormNumberInput
            type='number'
            label='Min amount'
            placeholder='Ex: -45 for -45€'
            value={minPrice}
            onChange={this.onPriceMinChanged}
          />
          <FormNumberInput arabic
            type='number'
            label='Max amount'
            placeholder='Ex: 45 for 45€'
            value={maxPrice}
            onChange={this.onPriceMaxChanged}
          />
        </Block>
        <FormInput
          type='text'
          label='Name'
          value={name}
          onChange={this.onNameChanged}
        />
        <FormSelector
          label='Category'
          color='primary'
          // style={{ width: '80%' }}
          values={this.categories}
          onChange={this.onCategorySelected}
        />
        <br />
        <ThemedButton bgcolor='primary' onClick={this.onSubmit}>Filtrer</ThemedButton>
        {hasError && <StyledError>{errors.global}</StyledError>}
      </Form>
    );
  }
}

export default FilterForm;
