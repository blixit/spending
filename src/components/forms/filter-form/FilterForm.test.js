import { withTheme } from 'ui/theming/test.utils';

import render from 'testing/render';

import FilterForm from './FilterForm';

describe('[components/forms/filter-form] FilterForm', () => {
  it('should render properly', () => {
    const filterForm = render(withTheme(FilterForm, {}));
    expect(filterForm.toJSON()).toMatchSnapshot();
  });
});
