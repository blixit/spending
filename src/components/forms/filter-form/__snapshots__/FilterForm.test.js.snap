// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[components/forms/filter-form] FilterForm should render properly 1`] = `
<form
  className="sc-jTzLTM iNxaQE"
>
  <div
    className="sc-jzJRlG fKTnNB"
  >
    <div
      className="sc-iwsKbI LIYKF"
    >
      <label
        className="sc-ifAKCX hbDuiX sc-dnqmqq kYHIQY"
      >
        <span>
          From
        </span>
        <div
          className="react-datepicker-wrapper"
        >
          <div
            className="react-datepicker__input-container"
          >
            <input
              className="sc-gzVnrw koFpYU"
              disabled={false}
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              placeholder="select a date"
              readOnly={false}
              type="text"
              value="18/09/2019"
            />
          </div>
        </div>
      </label>
    </div>
    <div
      className="sc-iwsKbI LIYKF"
    >
      <label
        className="sc-ifAKCX gtpqFH sc-dnqmqq dsKfiS"
      >
        <span>
          To
        </span>
        <div
          className="react-datepicker-wrapper"
        >
          <div
            className="react-datepicker__input-container"
          >
            <input
              className="sc-gzVnrw koFpYU"
              disabled={false}
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              placeholder="select a date"
              readOnly={false}
              type="text"
              value="18/10/2019"
            />
          </div>
        </div>
      </label>
    </div>
  </div>
  <div
    className="sc-jzJRlG fKTnNB"
  >
    <div
      className="sc-bZQynM funnpa sc-fjdhpX gXwiOt"
      placeholder="Ex: -45 for -45€"
    >
      <label
        className="sc-ifAKCX hbDuiX"
      >
        <span>
          Min amount
        </span>
        <input
          className="sc-bxivhb dwfPSM sc-fjdhpX gXwiOt"
          onChange={[Function]}
          placeholder="Ex: -45 for -45€"
          type="number"
        />
      </label>
    </div>
    <div
      className="sc-bZQynM funnpa sc-fjdhpX gXwiOt"
      placeholder="Ex: 45 for 45€"
    >
      <label
        className="sc-ifAKCX gtpqFH"
      >
        <span>
          Max amount
        </span>
        <input
          className="sc-bxivhb uZeYE sc-fjdhpX gXwiOt"
          onChange={[Function]}
          placeholder="Ex: 45 for 45€"
          type="number"
        />
      </label>
    </div>
  </div>
  <div
    className="sc-bZQynM funnpa"
  >
    <label
      className="sc-ifAKCX hbDuiX"
    >
      <span>
        Name
      </span>
      <input
        className="sc-bxivhb dwfPSM"
        onChange={[Function]}
        type="text"
      />
    </label>
  </div>
  <div
    className="sc-VigVT ddMFQP"
    color="primary"
  >
    <label
      className="sc-ifAKCX hbDuiX"
    >
      <span>
        Category
      </span>
      <select
        className="sc-gZMcBi hWLNG"
        color="primary"
        onChange={[Function]}
      >
        <option>
          Restauration
        </option>
        <option>
          Taxis
        </option>
      </select>
    </label>
  </div>
  <br />
  <button
    className="sc-bwzfXH cuSBsa sc-htpNat knmdAQ"
    onClick={[Function]}
  >
    Filtrer
  </button>
</form>
`;
